### PR TITLE
[BEAM-2879] Support writing data to BigQuery via avro

### PR DESCRIPTION
**Pull request by steveniemitz**
This change enhances BigQueryIO.Write to support writing avro files rather than json when using FILE_LOADS (STREAMING_INSERTS is unchanged).

### Benchmarks

Preliminary results look good.  The more CPU constrained a job is, the faster avro becomes.

My test dataset is a typical workload of ours, around 2 billion records (~130 GB serialized) representing the result of a combine.  My tests read these records from GCS and wrote them to BigQuery.  The jobs were run in dataflow with 150 x n1-standard-2 workers.

format | time to start load job | bytes written | BQ slot time (ms)
-------|----------------------|--------------|-------------
avro    |  6 m 30 s                     | 126 GB          | 35,189,679
json    |  8 m 5 s                        |  712 GB         | 96,006,088

@lukecwik @chamikaramj looks like you guys are the OWNERS of the gcp IO module.

------------------------

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by pabloem** [comment]
This is just a brain dump of what I'm thinking...

I wonder whether we need the `AvroWriteRequest`, and the Avro schema. I guess we do, as the `InputElement` (whatever it is) + the Avro schema are all one needs to build the `GenericRecord`. Having the `AvroWriteRequest` may help make the formatting function as concise as possible....

As for supporting Beam schemas + avro files, one could have a `useBeamSchemaForAvroFiles()`... though it's a little strange....

Another option is to have `useBeamSchema`, and a pre-coded avro formatting function called something like ... `BigQueryIOUtils.beamRowToAvroRecord()`. Though this is a little awkward too.

---
Overall, I like using `AvroWriteRequest` as input for the avro format function... and for supporting Beam schemas, it may be that `useBeamSchemaForAvroFiles` (or some better name) is the more reasonable options.
WDYT?

**Comment by steveniemitz** [comment]
Thanks for the thoughts!  My comments inline

> This is just a brain dump of what I'm thinking...
> 
> I wonder whether we need the `AvroWriteRequest`, and the Avro schema. I guess we do, as the `InputElement` (whatever it is) + the Avro schema are all one needs to build the `GenericRecord`. Having the `AvroWriteRequest` may help make the formatting function as concise as possible....

I really went back and forth on this a few times.  We could use `SerializableBiFunction` here, but if in the future we ever wanted to add another parameter, it'd be a breaking change.  This was we can just add a field to the class.  This follows the same pattern as read does, where it takes a `SchemaAndRecord` as an input.  You do need both the avro schema and the element though in order to support more advanced cases w/ DynamicDestinations, etc.  Plus avro schemas themselves aren't easily serializable (until avro 1.9) so users can't simply create a closure over them.

I do hate the name though, if you can think of anything better I'd love to rename this!
 
> As for supporting Beam schemas + avro files, one could have a `useBeamSchemaForAvroFiles()`... though it's a little strange....
> 
> Another option is to have `useBeamSchema`, and a pre-coded avro formatting function called something like ... `BigQueryIOUtils.beamRowToAvroRecord()`. Though this is a little awkward too.

Yeah I struggled with this as well.  The only thing stopping us from having a version that supports beam schemas is the interface.  `useBeamSchemaForAvroFiles` is a pretty reasonable name. 

> Overall, I like using `AvroWriteRequest` as input for the avro format function... and for supporting Beam schemas, it may be that `useBeamSchemaForAvroFiles` (or some better name) is the more reasonable options.
> WDYT?

 I'd be up for adding that in a follow-up PR.  I also have some ideas around `writeGenericRecords()` I want to play around with (that would also use beam schemas, similar to `AvroIO.writeGenericRecords()` )

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtilsTest.java [hunk 1]
```diff
@@ -243,8 +243,8 @@ public void testConvertBigQuerySchemaToAvroSchema() {
                 Schema.create(Type.NULL),
                 Schema.createRecord(
                     "scion",
-                    "org.apache.beam.sdk.io.gcp.bigquery",
                     "Translated Avro Schema for scion",
+                    "org.apache.beam.sdk.io.gcp.bigquery",
```
[the remark this answers is not in the document]
Comment (steveniemitz) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtilsTest.java]: oh no actually, this is pretty important!  The previous version flipped the namespace and description parameters, resulting in an invalid namespace.  See the corresponding change up in BigQueryAvroUtils.

Really avro's fault for making a function with 3 string parameters...

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AbstractRowWriter.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import java.nio.channels.WritableByteChannel;
+import java.util.UUID;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.io.CountingOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Writes {@link TableRow} objects out to a file. Used when doing batch load jobs into BigQuery. */
+abstract class AbstractRowWriter<T> implements AutoCloseable {
```
Comment (steveniemitz) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AbstractRowWriter.java]: nit: I don't usually write java, is the style convention `AbstractDerp` for base classes like this?  or should this just be `RowWriter<T>`?

Comment (pabloem) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AbstractRowWriter.java]: In my experience, the Abstract is usually not added to the abstract class. I am okay with either.

**Comment by pabloem** [comment]
LMK if you'd like me to take another look.

**Comment by steveniemitz** [comment]
> LMK if you'd like me to take another look.

oh, yeah please do.  I don't have much more from my end other than renaming the class above.

**Comment by pabloem** [comment]
Run Java PreCommit

**Comment by steveniemitz** [comment]
Thanks!  I'll do the class rename asap too.

**Comment by steveniemitz** [comment]
Thanks!  You can merge this now and I'll do the rename in another PR, or I can get to it tomorrow, up to you!

**Comment by pabloem** [comment]
I don't think we strictly need the rename. But I do want to wait for @chamikaramj to take a look : ) - I don't think he'll have objections, but just in case he thinks of any improvements.

**Comment by chamikaramj** [comment]
Sorry about the delay. Taking a look.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AbstractRowWriter.java (hunk 2 above)
Comment (chamikaramj) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AbstractRowWriter.java]: +1 for just RowWriter or BigQueryRowWriter to be more specific.

**Comment by steveniemitz** [comment]
> Thanks. Looks great to me.
> 
> To make sure I understood correctly, this will not be enabled for existing users by default and to enable this users have to specify withAvroFormatFunction(), correct ?
 
Correct, with schemas I think we could make this enabled transparently, but for now its opt-in only.

> Also, can we add a version of BigQueryIOIT so that we can continue to monitor both Avro and JSON based BQ write transforms ?

Yeah I can add that in there.

**Comment by steveniemitz** [comment]
@chamikaramj are there up-to-date docs anywhere on how to run the BigQueryIOIT tests?  The example args in the javadoc are super out of date.

**Comment by steveniemitz** [comment]
> @chamikaramj are there up-to-date docs anywhere on how to run the BigQueryIOIT tests? The example args in the javadoc are super out of date.

nm I hacked it up enough to get it to run.

**Comment by chamikaramj** [comment]
Thanks. 

Do TODOs in PR description still apply ?

**Comment by chamikaramj** [comment]
Run Java PostCommit

**Comment by chamikaramj** [comment]
Also please fixup commits before merging.

**Comment by steveniemitz** [comment]
> Thanks.
> 
> Do TODOs in PR description still apply ?

nope, I just edited the message

**Comment by steveniemitz** [comment]
> Also please fixup commits before merging.

Was this for me or whoever ends up merging it?

**Comment by chamikaramj** [comment]
It's for you :)

**Comment by steveniemitz** [comment]
coolio, fwiw, the contribution guide is ambiguous wrt who should do the squashing.

> Beam committers can squash all commits in the PR during merge, however if a PR has a mixture of independent changes that should not be squashed, and fixup commits, then the PR author should help squashing fixup commits to maintain a clean commmit history.

**Comment by chamikaramj** [comment]
Thanks. LGTM.

Yeah, committer can squash and commit if you just need one commit.

**Comment by chamikaramj** [comment]
Run Java PostCommit

**Comment by chamikaramj** [comment]
Will merge after post-commit tests pass.

**Comment by chamikaramj** [comment]
Unfortunately, looks like BigQueryIOIT is a recently added test that is currently not captured by any of the test suites.

@steveniemitz Will you be able to add a test that writes to BQ using Avro to the BigQueryTornadoesIT that is captured by the Beam Java PostCommit test suite ?
https://github.com/apache/beam/blob/master/examples/java/src/test/java/org/apache/beam/examples/cookbook/BigQueryTornadoesIT.java
https://builds.apache.org/view/A-D/view/Beam/view/PostCommit/job/beam_PostCommit_Java/lastCompletedBuild/testReport/org.apache.beam.examples.cookbook/BigQueryTornadoesIT/

@mwalenia can you comment on the status of BigQueryIOIT ?

cc: @pabloem

**Comment by steveniemitz** [comment]
> @steveniemitz Will you be able to add a test that writes to BQ using Avro to the BigQueryTornadoesIT that is captured by the Beam Java PostCommit test suite ?

honestly at this point I'm not going to prioritize doing so.

**Comment by chamikaramj** [comment]
Ok, fair enough.

I'll go ahead and merge this. But please consider adding a regularly running integration test in a follow up PR to make sure that this codepath does not become stale/broken.

**Comment by steveniemitz** [comment]
maybe we should just make BigQueryIOIT actually run ;)

**Pull request merged**